README.md
# Features
It allows to send all JMeter data to SignalFX and also report events about start and end of tests

This listener can be used with JMeter 3.1

# Parameters
Parameters allows you to specify where to send metrics and what to actually send

| Parameter | Description | Values |  Default Value |
|---|---|---|---| 
| signalFxEndpoint|Endpoint where to send metrics| | https://ingest.signalfx.com:443 |
| authToken	| Auth token to use when sending metrics. Get one in SignalFX web app	| auth token string ||	
|dimensions	|SignalFX dimensions.	|Semicolon separated list of key=value pairs	|app=jmeter|
|summaryOnly|	Should only "all" samplers be reported|	true/false|	false|
|samplersList|	List of samplers that should be reported (or regex)|	Semicolon separated list of samplers or regex (if useRegexpForSamplersList is true)|	.*|
|useRegexpForSamplersList|	Regex to match reported samplers |	true/false	| true |
|percentiles|	Percentiles that should be reported | 	Semicolon separated list of percentile values	| 90;95;99|
|testTitle|	Title for event started, recorded as property	| String that will help you to identify your test in signal fx events	| Test name |
|eventTags|	Dimensions for "events" about start and end of test	|Semicolon separated list of key=value pairs ||	

Example parameters equal to what was used to send metrics to Influx/Grafana

# Example of dimensions and event titles
|Parameter| Value | 
|---|---|
|dimensions|	app=backend;project=project_x;env=pre-prod;app_loc=usw2;gen_loc=use1 |
|testTitle|	Project X: Load round one |

# Installation

1. `mvn clean package`
2. Copy `target/signalfx-listener.jar` to `lib/ext` folder of your JMeter installation
3. After that you can add backend listener to your project and select SignalFX in combobox

## Note

JMeter 3.1 does not "memoize" settings when you switch one listener to another, so be careful with 
switching from Graphite to SignalFX and back

CHANGELOG.md
## 0.1
  - Initial release of plugin

src/main/java/com/aligntech/cs/jmeter/signalfx/SamplersFilter.java
/*
 * Copyright 2017 Align Technology, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aligntech.cs.jmeter.signalfx;

public interface SamplersFilter {

    /**
     * Test if sampler should be recorded (sent to SFX) or not
     * @param sampler sampler name
     * @return {@code true} if sampler data should be recorded and {@code false} otherwise
     */
    boolean metricReported(String sampler);

    static SamplersFilter create(String samplers, boolean useRegex) {
        if (useRegex) {
            return new RegexSamplersFilter(samplers);
        } else {
            return new ListSamplersFilter(samplers);
        }
    }

}


src/main/java/com/aligntech/cs/jmeter/signalfx/MetricCollector.java
/*
Copyright 2017 Align Technology

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */


package com.aligntech.cs.jmeter.signalfx;

import javax.annotation.concurrent.NotThreadSafe;

/**
 * Collects metrics of different kinds. Metrics are not automatically sent. To send metrics collector should
 * be passed to proper metric sender.
 */
@NotThreadSafe
public interface MetricCollector {
    /**
     * Record gauge (instant value) for given measurement and context
     * @param contextName section where measurement is done (e.g. test)
     * @param measurement kind of measurement (e.g. minAT)
     * @param value integer metric value
     */
    void addGauge(String contextName, String measurement, int value);

    /**
     * Record gauge (instant value) for given measurement and context
     * @param contextName section where measurement is done (e.g. all, specific measurement)
     * @param measurement kind of measurement (e.g. ok.avg)
     * @param value double metric value
     */
    void addGauge(String contextName, String measurement, double value);

    /**
     * Add value to a counter defined by given context and measurement
     * @param contextName section where measurement is done (e.g. all, specific measurement)
     * @param measurement kind of measurement (e.g. ok.avg)
     * @param value value to be added to counter
     */
    void addCounter(String contextName, String measurement, int value);
}


src/main/java/com/aligntech/cs/jmeter/signalfx/ListSamplersFilter.java
/*
 * Copyright 2017 Align Technology, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aligntech.cs.jmeter.signalfx;

import com.google.common.collect.ImmutableSet;

import java.util.Set;

/**
 * Simple filter that does exact match of sampler against semicolon separated list of samples
 */
class ListSamplersFilter implements SamplersFilter {

    private final Set<String> samplers;

    public ListSamplersFilter(String samplerString) {
        this.samplers = ImmutableSet.copyOf(samplerString.split(";"));
    }

    @Override
    public boolean metricReported(String sampler) {
        return samplers.contains(sampler);
    }
}


src/main/java/com/aligntech/cs/jmeter/signalfx/RegexSamplersFilter.java
/*
 * Copyright 2017 Align Technology, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aligntech.cs.jmeter.signalfx;

import java.util.regex.Pattern;

/**
 * Regular expression based filter. All string is considered as regular expression that is used to match
 * against sampler value
 */
class RegexSamplersFilter implements SamplersFilter {

    private final Pattern pattern;

    RegexSamplersFilter(String samplers) {
        pattern = Pattern.compile(samplers);
    }

    @Override
    public boolean metricReported(String sampler) {
        return pattern.matcher(sampler).matches();
    }
}


src/main/java/com/aligntech/cs/jmeter/signalfx/SignalFxMetricsSender.java
/*
 * Copyright 1998-2017 The Apache Software Foundation
 * Copyright 2017 Align Technology, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aligntech.cs.jmeter.signalfx;

/**
 * Abstraction to send metrics to SignalFx similar to Graphite and SignalFx hosted in JMeter codebase
 */
interface SignalFxMetricsSender {

    /**
     * Record and send event related to test run
     * @param event event type
     * @param details description of event
     */
    void addEvent(String event, String details);

    /**
     * Creates collector to do record one set of metrics. Returned collector is <strong>not</strong> threadsafe
     * and should not be shared by multiple threads
     */
    MetricCollector startCollection();

    /**
     * Sends all metrics collected by given collector and closes it for further usage
     */
    void writeAndSendMetrics(MetricCollector collector);

    /**
     * Closes all resources associated with this sender
     */
    void destroy();

}


src/main/java/com/aligntech/cs/jmeter/signalfx/SignalFxMetricsSenderImpl.java
/*
 * Copyright 2017 Align Technology, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aligntech.cs.jmeter.signalfx;

import com.google.common.base.Throwables;
import com.signalfx.endpoint.SignalFxEndpoint;
import com.signalfx.metrics.auth.StaticAuthToken;
import com.signalfx.metrics.connection.HttpDataPointProtobufReceiverFactory;
import com.signalfx.metrics.connection.HttpEventProtobufReceiverFactory;
import com.signalfx.metrics.errorhandler.MetricError;
import com.signalfx.metrics.errorhandler.MetricErrorType;
import com.signalfx.metrics.errorhandler.OnSendErrorHandler;
import com.signalfx.metrics.flush.AggregateMetricSender;
import com.signalfx.metrics.protobuf.SignalFxProtocolBuffers;
import org.apache.jorphan.logging.LoggingManager;
import org.apache.log.Logger;

import java.net.URI;
import java.util.*;

import static com.signalfx.metrics.errorhandler.MetricErrorType.DATAPOINT_SEND_ERROR;
import static com.signalfx.metrics.errorhandler.MetricErrorType.EVENT_SEND_ERROR;

public class SignalFxMetricsSenderImpl implements SignalFxMetricsSender, OnSendErrorHandler {

    private static final Logger log = LoggingManager.getLoggerForClass();

    private final AggregateMetricSender sender;
    private final List<SignalFxProtocolBuffers.Dimension> dimensions;
    private final Map<String, String> eventDimensions;

    public SignalFxMetricsSenderImpl(String endpoint, String authToken,
                                     Map<String, String> dimensions,
                                     Map<String, String> eventDimensions, int timeout) {
        URI uri = toUri(endpoint);

        sender = new AggregateMetricSender("jmeter",
                new HttpDataPointProtobufReceiverFactory(new SignalFxEndpoint(uri.getScheme(), uri.getHost(), uri.getPort())).setTimeoutMs(timeout),
                new HttpEventProtobufReceiverFactory(new SignalFxEndpoint(uri.getScheme(), uri.getHost(), uri.getPort())).setTimeoutMs(timeout),
                new StaticAuthToken(authToken), Collections.singletonList(this));
        this.dimensions = new ArrayList<>(dimensions.size());
        for (Map.Entry<String, String> d : dimensions.entrySet()) {
            this.dimensions.add(dimension(d).build());
        }
        this.eventDimensions = eventDimensions;
    }

    private URI toUri(String endpoint) {
        try {
            return new URI(endpoint);
        } catch (Exception ex) {
            throw Throwables.propagate(ex);
        }
    }

    @Override
    public void addEvent(String event, String details) {
        AggregateMetricSender.Session session = sender.createSession();
        SignalFxProtocolBuffers.Event.Builder builder = SignalFxProtocolBuffers.Event.newBuilder()
                .setEventType("Jmeter: " + event)
                .setCategory(SignalFxProtocolBuffers.EventCategory.USER_DEFINED)
                .setTimestamp(System.currentTimeMillis())
                .addDimensions(
                        dimension("source", "jmeter"))
                .addProperties(
                        SignalFxProtocolBuffers.Property.newBuilder()
                                .setKey("details")
                                .setValue(
                                        SignalFxProtocolBuffers.PropertyValue.newBuilder()
                                                .setStrValue(details)
                                                .build())
                                .build());
        for (Map.Entry<String, String> eventDimensions : eventDimensions.entrySet()) {
            builder.addDimensions(dimension(eventDimensions));
        }
        session.setEvent(builder.build());
        closeQuitely(session);
    }

    private SignalFxProtocolBuffers.Dimension.Builder dimension(Map.Entry<String, String> entry) {
        return dimension(entry.getKey(), entry.getValue());
    }

    private SignalFxProtocolBuffers.Dimension.Builder dimension(String key, String value) {
        return SignalFxProtocolBuffers.Dimension.newBuilder().setKey(key).setValue(value);
    }

    @Override
    public MetricCollector startCollection() {
        return new MetricCollectorImpl(sender.createSession(), dimensions);
    }

    @Override
    public void writeAndSendMetrics(MetricCollector collector) {
        if (!(collector instanceof MetricCollectorImpl)) {
            throw new IllegalArgumentException("This sender can work only with " + MetricCollectorImpl.class);
        }
        closeQuitely(((MetricCollectorImpl) collector).session);
    }

    private void closeQuitely(AggregateMetricSender.Session session) {
        try {
            session.close();
        } catch (Exception ex) {
            //to be handled in handleError(MetricError) method
        }
    }

    @Override
    public void destroy() {
        log.debug("Destroing signalfx sender");
    }

    @Override
    public void handleError(MetricError metricError) {
        Set<MetricErrorType> warnErrors = EnumSet.of(DATAPOINT_SEND_ERROR,
                EVENT_SEND_ERROR);
        if (warnErrors.contains(metricError.getMetricErrorType())) {
            log.warn(metricError.getMessage(), metricError.getException());
        } else {
            log.error(metricError.getMessage(), metricError.getException());
        }
    }

    private static class MetricCollectorImpl implements MetricCollector {

        private final long timestamp;
        private final AggregateMetricSender.Session session;
        private final List<SignalFxProtocolBuffers.Dimension> dimensions;

        public MetricCollectorImpl(AggregateMetricSender.Session session, List<SignalFxProtocolBuffers.Dimension> dimensions) {
            this.session = session;
            this.timestamp = System.currentTimeMillis();
            this.dimensions = dimensions;
        }

        @Override
        public void addGauge(String contextName, String measurement, int value) {
            session.setDatapoint(metric(contextName, measurement)
                    .setMetricType(SignalFxProtocolBuffers.MetricType.GAUGE)
                    .setValue(SignalFxProtocolBuffers.Datum.newBuilder().setIntValue(value))
                    .build()
            );
        }

        @Override
        public void addGauge(String contextName, String measurement, double value) {
            if (Double.isInfinite(value) || Double.isNaN(value)) {
                return;
            }
            session.setDatapoint(metric(contextName, measurement)
                    .setMetricType(SignalFxProtocolBuffers.MetricType.GAUGE)
                    .setValue(SignalFxProtocolBuffers.Datum.newBuilder().setDoubleValue(value))
                    .build()
            );
        }

        @Override
        public void addCounter(String contextName, String measurement, int value) {
            session.setDatapoint(metric(contextName, measurement)
                    .setMetricType(SignalFxProtocolBuffers.MetricType.COUNTER)
                    .setValue(SignalFxProtocolBuffers.Datum.newBuilder().setIntValue(normalizeValue(value)))
                    .build()
            );
        }

        /**
         * Normalization of integer value. For some reason passing of value without multiplication results in
         * strange graphed values
         */
        private int normalizeValue(int value) {
            return value * 10;
        }

        private SignalFxProtocolBuffers.DataPoint.Builder metric(String contextName, String measurement) {
            return SignalFxProtocolBuffers.DataPoint.newBuilder()
                    .setTimestamp(timestamp)
                    .setSource("jmeter")
                    .setMetric("jmeter." + contextName + "." + measurement)
                    .addAllDimensions(dimensions);
        }
    }
}


src/main/java/com/aligntech/cs/jmeter/signalfx/SignalFxBackendListenerClient.java
/*
 * Copyright 1998-2017 The Apache Software Foundation
 * Copyright 2017 Align Technology, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aligntech.cs.jmeter.signalfx;

import com.google.common.collect.ImmutableMap;
import com.signalfx.endpoint.SignalFxEndpoint;
import com.signalfx.metrics.connection.HttpDataPointProtobufReceiverFactory;
import org.apache.commons.lang3.StringUtils;
import org.apache.jmeter.config.Arguments;
import org.apache.jmeter.samplers.SampleResult;
import org.apache.jmeter.visualizers.backend.AbstractBackendListenerClient;
import org.apache.jmeter.visualizers.backend.BackendListenerContext;
import org.apache.jmeter.visualizers.backend.SamplerMetric;
import org.apache.jmeter.visualizers.backend.UserMetric;
import org.apache.jorphan.logging.LoggingManager;
import org.apache.log.Logger;

import java.text.DecimalFormat;
import java.util.*;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Backend listener implementation that records and sends data to SignalFX service.
 *
 * Note: BackendListener interface itself is too abstract to create code devoted solely for working with SFX, that's
 * why boilerplate code in this class related to creating list of properties and iterating over metrics is taken from
 * GraphiteBackendListener implementation and adapted to SFX needs with slight refactoring.
 */
public class SignalFxBackendListenerClient extends AbstractBackendListenerClient implements Runnable {

    private static final String SIGNAL_FX_DEFAULT_ENDPOINT = SignalFxEndpoint.DEFAULT_SCHEME + "://" + SignalFxEndpoint.DEFAULT_HOSTNAME + ":" + SignalFxEndpoint.DEFAULT_PORT;
    private static final Logger log = LoggingManager.getLoggerForClass();

    private static final String SEPARATOR = ";";

    private static final String CUMULATED_METRICS = "_all_";

    private static final String METRIC_OK_COUNT = "ok.count";
    private static final String METRIC_OK_MIN_RESPONSE_TIME = "ok.min";
    private static final String METRIC_OK_MAX_RESPONSE_TIME = "ok.max";
    private static final String METRIC_OK_AVG_RESPONSE_TIME = "ok.avg";

    private static final String METRIC_KO_COUNT = "ko.count";
    private static final String METRIC_KO_MIN_RESPONSE_TIME = "ko.min";
    private static final String METRIC_KO_MAX_RESPONSE_TIME = "ko.max";
    private static final String METRIC_KO_AVG_RESPONSE_TIME = "ko.avg";

    private static final String METRIC_ALL_COUNT = "a.count";
    private static final String METRIC_ALL_MIN_RESPONSE_TIME = "a.min";
    private static final String METRIC_ALL_MAX_RESPONSE_TIME = "a.max";
    private static final String METRIC_ALL_AVG_RESPONSE_TIME = "a.avg";

    private static final String METRIC_ALL_HITS_COUNT = "h.count";

    // User Metrics
    private static final String METRIC_MAX_ACTIVE_THREADS = "maxAT";
    private static final String METRIC_MIN_ACTIVE_THREADS = "minAT";
    private static final String METRIC_MEAN_ACTIVE_THREADS = "meanAT";
    private static final String METRIC_STARTED_THREADS = "startedT";
    private static final String METRIC_FINISHED_THREADS = "endedT";

    private static final String SIGNAL_FX_ENDPOINT = "signalFxEndpoint";
    private static final String AUTH_TOKEN = "authToken";
    private static final String DIMENSIONS = "dimensions";
    private static final String SAMPLERS_LIST = "samplersList";
    private static final String USE_REGEXP_FOR_SAMPLERS_LIST = "useRegexpForSamplersList";
    private static final String PERCENTILES = "percentiles";
    private static final String SIGNALFX_TIMEOUT = "signalFxTimeoutMs";

    private static final String TEST_TITLE = "testTitle";
    private static final String TEST_TITLE_DEFAULT = "Test name";

    private static final String EVENT_TAGS = "eventTags";
    private static final String SUMMARY_ONLY = "summaryOnly";

    private static final long METRIC_SEND_INTERVAL = 5L;
    private static final String PERCENTILES_DEFAULT = "90;95;99";

    private static final String ALL_CONTEXT_NAME = "all";
    private static final String TEST_CONTEXT_NAME = "test";

    private Map<String, Float> okPercentiles;
    private Map<String, Float> koPercentiles;
    private Map<String, Float> allPercentiles;

    private ScheduledExecutorService scheduler;
    private ScheduledFuture timerHandle;

    private boolean summaryOnly;
    private String testTitle;
    private SamplersFilter filter;

    private SignalFxMetricsSender metricsSender;

    private Lock lock = new ReentrantLock();

    @Override
    public void run() {
        sendMetrics();
    }

    private void sendMetrics() {
        MetricCollector metricCollector = metricsSender.startCollection();
        try {
            lock.lock();
            for (Map.Entry<String, SamplerMetric> entry : getMetricsPerSampler().entrySet()) {
                final String key = entry.getKey();
                final SamplerMetric metric = entry.getValue();
                if (key.equals(CUMULATED_METRICS)) {
                    addMetrics(metricCollector, ALL_CONTEXT_NAME, metric);
                } else {
                    addMetrics(metricCollector, key, metric);
                }
                // We are computing on interval basis so cleanup
                metric.resetForTimeInterval();
            }
        } finally {
            lock.unlock();
        }
        UserMetric userMetric = getUserMetrics();
        metricCollector.addGauge(TEST_CONTEXT_NAME, METRIC_MIN_ACTIVE_THREADS, userMetric.getMinActiveThreads());
        metricCollector.addGauge(TEST_CONTEXT_NAME, METRIC_MAX_ACTIVE_THREADS, userMetric.getMaxActiveThreads());
        metricCollector.addGauge(TEST_CONTEXT_NAME, METRIC_MEAN_ACTIVE_THREADS, userMetric.getMeanActiveThreads());
        metricCollector.addGauge(TEST_CONTEXT_NAME, METRIC_STARTED_THREADS, userMetric.getStartedThreads());
        metricCollector.addGauge(TEST_CONTEXT_NAME, METRIC_FINISHED_THREADS, userMetric.getFinishedThreads());

        metricsSender.writeAndSendMetrics(metricCollector);
    }

    private void addMetrics(MetricCollector collector, String contextName, SamplerMetric metric) {
        if (metric.getTotal() > 0) {
            collector.addCounter(contextName, METRIC_OK_COUNT, metric.getSuccesses());
            collector.addCounter(contextName, METRIC_KO_COUNT, metric.getFailures());
            collector.addCounter(contextName, METRIC_ALL_COUNT, metric.getTotal());
            collector.addCounter(contextName, METRIC_ALL_HITS_COUNT, metric.getHits());
            if (metric.getSuccesses() > 0) {
                collector.addGauge(contextName, METRIC_OK_MIN_RESPONSE_TIME, metric.getOkMinTime());
                collector.addGauge(contextName, METRIC_OK_MAX_RESPONSE_TIME, metric.getOkMaxTime());
                collector.addGauge(contextName, METRIC_OK_AVG_RESPONSE_TIME, metric.getOkMean());
                for (Map.Entry<String, Float> entry : okPercentiles.entrySet()) {
                    collector.addGauge(contextName, entry.getKey(),
                            metric.getOkPercentile(entry.getValue().doubleValue()));
                }
            }
            if (metric.getFailures() > 0) {
                collector.addGauge(contextName, METRIC_KO_MIN_RESPONSE_TIME, metric.getKoMinTime());
                collector.addGauge(contextName, METRIC_KO_MAX_RESPONSE_TIME, metric.getKoMaxTime());
                collector.addGauge(contextName, METRIC_KO_AVG_RESPONSE_TIME, metric.getKoMean());
                for (Map.Entry<String, Float> entry : koPercentiles.entrySet()) {
                    collector.addGauge(contextName, entry.getKey(),
                            metric.getKoPercentile(entry.getValue().doubleValue()));
                }
            }
            collector.addGauge(contextName, METRIC_ALL_MIN_RESPONSE_TIME, metric.getAllMinTime());
            collector.addGauge(contextName, METRIC_ALL_MAX_RESPONSE_TIME, metric.getAllMaxTime());
            collector.addGauge(contextName, METRIC_ALL_AVG_RESPONSE_TIME, metric.getAllMean());
            for (Map.Entry<String, Float> entry : allPercentiles.entrySet()) {
                collector.addGauge(contextName, entry.getKey(),
                        metric.getAllPercentile(entry.getValue().doubleValue()));
            }
        }
    }

    @Override
    public void handleSampleResults(List<SampleResult> sampleResults, BackendListenerContext backendListenerContext) {
        try {
            lock.lock();
            UserMetric userMetrics = getUserMetrics();
            for (SampleResult sampleResult : sampleResults) {
                userMetrics.add(sampleResult);

                if (!summaryOnly) {
                    if (filter.metricReported(sampleResult.getSampleLabel())) {
                        SamplerMetric samplerMetric = getSamplerMetric(sampleResult.getSampleLabel());
                        samplerMetric.add(sampleResult);
                    }
                }
                SamplerMetric cumulatedMetrics = getSamplerMetric(CUMULATED_METRICS);
                cumulatedMetrics.add(sampleResult);
            }
        } finally {
            lock.unlock();
        }
    }

    @Override
    public void setupTest(BackendListenerContext context) throws Exception {
        super.setupTest(context);
        initDefaults();

        summaryOnly = context.getBooleanParameter(SUMMARY_ONLY, false);
        filter = createFilter(context);
        testTitle = context.getParameter(TEST_TITLE, TEST_TITLE_DEFAULT);
        setupPercentiles(extractPercentiles(context.getParameter(PERCENTILES, "")));

        metricsSender = new SignalFxMetricsSenderImpl(
                context.getParameter(SIGNAL_FX_ENDPOINT, SIGNAL_FX_DEFAULT_ENDPOINT),
                context.getParameter(AUTH_TOKEN),
                parseDimensions(context.getParameter(DIMENSIONS, "")),
                parseDimensions(context.getParameter(EVENT_TAGS, "")),
                context.getIntParameter(SIGNALFX_TIMEOUT, HttpDataPointProtobufReceiverFactory.DEFAULT_TIMEOUT_MS)
        );

        reportTestsStart();

        this.scheduler = Executors.newScheduledThreadPool(1);
        this.timerHandle = this.scheduler.scheduleAtFixedRate(this, 1L, METRIC_SEND_INTERVAL, TimeUnit.SECONDS);
    }

    private void initDefaults() {
        this.okPercentiles = new HashMap<>();
        this.allPercentiles = new HashMap<>();
        this.koPercentiles = new HashMap<>();
    }

    private SamplersFilter createFilter(BackendListenerContext context) {
        String samplersList = context.getParameter(SAMPLERS_LIST, "");
        boolean useRegexForSamplers = context.getBooleanParameter(USE_REGEXP_FOR_SAMPLERS_LIST, false);
        return SamplersFilter.create(samplersList, useRegexForSamplers);
    }

    private Collection<String> extractPercentiles(String userValue) {
        String[] split = userValue.split(SEPARATOR);
        ArrayList<String> percentiles = new ArrayList<>();
        for (String value : split) {
            String trimmedValue = value.trim();
            if (!StringUtils.isEmpty(trimmedValue)) {
                percentiles.add(trimmedValue);
            }
        }
        return percentiles;
    }

    private void setupPercentiles(Collection<String> percentiles) {
        DecimalFormat format = new DecimalFormat("0.##");
        for (String percentile : percentiles) {
            try {
                Float percentileValue = Float.valueOf(percentile.trim());
                String formattedPercentile = format.format(percentileValue);
                okPercentiles.put(metricName("ok.pct" + formattedPercentile), percentileValue);
                koPercentiles.put(metricName("ko.pct" + formattedPercentile), percentileValue);
                allPercentiles.put(metricName("a.pct" + formattedPercentile), percentileValue);
            } catch (Exception e) {
                log.error("Error parsing percentile: " + percentile, e);
            }
        }

    }

    private String metricName(String name) {
        return name;
    }

    private Map<String, String> parseDimensions(String dimensionsString) {
        Map<String, String> dimensions = new LinkedHashMap<>();
        for (String dimensionString : dimensionsString.split(";")) {
            if (StringUtils.isEmpty(dimensionsString)) continue;

            if (dimensionString.contains("=")) {
                int index = dimensionString.indexOf('=');
                dimensions.put(dimensionString.substring(0, index), dimensionString.substring(index + 1));
            } else {
                log.warn("Dimension has invalid format: " + dimensionString);
            }
        }
        return ImmutableMap.copyOf(dimensions);
    }

    @Override
    public void teardownTest(BackendListenerContext context) throws Exception {
        boolean cancelState = this.timerHandle.cancel(false);
        log.debug("Canceled state:" + cancelState);

        this.scheduler.shutdown();

        try {
            this.scheduler.awaitTermination(30L, TimeUnit.SECONDS);
        } catch (InterruptedException ex) {
            log.error("Error waiting for end of scheduler", ex);
        }

        log.info("Sending last metrics");
        this.sendMetrics();

        reportTestsEnd();

        this.filter = null;
        this.metricsSender.destroy();

        super.teardownTest(context);
    }

    private void reportTestsStart() {
        metricsSender.addEvent("started", testTitle);
    }

    private void reportTestsEnd() {
        metricsSender.addEvent("ended", testTitle);
    }

    @Override
    public Arguments getDefaultParameters() {
        Arguments arguments = new Arguments();
        arguments.addArgument(SIGNAL_FX_ENDPOINT, SIGNAL_FX_DEFAULT_ENDPOINT);
        arguments.addArgument(AUTH_TOKEN, "");
        arguments.addArgument(DIMENSIONS, "app=jmeter");
        arguments.addArgument(SUMMARY_ONLY, "false");
        arguments.addArgument(SAMPLERS_LIST, ".*");
        arguments.addArgument(USE_REGEXP_FOR_SAMPLERS_LIST, "true");
        arguments.addArgument(PERCENTILES, PERCENTILES_DEFAULT);
        arguments.addArgument(TEST_TITLE, TEST_TITLE_DEFAULT);
        arguments.addArgument(EVENT_TAGS, "");
        arguments.addArgument(SIGNALFX_TIMEOUT, Integer.toString(HttpDataPointProtobufReceiverFactory.DEFAULT_TIMEOUT_MS));
        return arguments;
    }
}
